src/backend/winit.rs

```diff
@@ -0,0 +1,652 @@
+    time_counter: u32,
+    key_counter: u32,
+    seat: Seat,
+    input_config: (),
+    handler: Option<Box<InputHandler<WinitInputBackend> + 'static>>,
+}
+
+/// Create a new `WinitGraphicsBackend`, which implements the `EGLGraphicsBackend`
+/// graphics backend trait and a corresponding `WinitInputBackend`, which implements
+/// the `InputBackend` trait
+pub fn init() -> Result<(WinitGraphicsBackend, WinitInputBackend), CreationError> {
+    init_from_builder(WindowBuilder::new())
+}
+
+/// Create a new `WinitGraphicsBackend`, which implements the `EGLGraphicsBackend`
+/// graphics backend trait, from a given `WindowBuilder` struct and a corresponding
+/// `WinitInputBackend`, which implements the `InputBackend` trait
+pub fn init_from_builder(builder: WindowBuilder)
+                         -> Result<(WinitGraphicsBackend, WinitInputBackend), CreationError> {
+    init_from_builder_with_gl_attr(builder,
+                                   GlAttributes {
+                                       version: None,
+                                       profile: None,
+                                       debug: cfg!(debug_assertions),
+                                       robustness: Robustness::TryRobustLoseContextOnReset,
+                                       vsync: true,
+                                   })
+}
+
+/// Create a new `WinitGraphicsBackend`, which implements the `EGLGraphicsBackend`
+/// graphics backend trait, from a given `WindowBuilder` struct, as well as given
+/// `GlAttributes` for further customization of the rendering pipeline and a
+/// corresponding `WinitInputBackend`, which implements the `InputBackend` trait.
+pub fn init_from_builder_with_gl_attr(builder: WindowBuilder, attributes: GlAttributes)
+                                      -> Result<(WinitGraphicsBackend, WinitInputBackend), CreationError> {
+    let events_loop = EventsLoop::new();
+    let window = Rc::new(builder.build(&events_loop)?);
+
+    let (native, surface) = if let (Some(conn), Some(window)) =
+        (get_x11_xconnection(), window.get_xlib_window()) {
+        (Native::X11(conn.display as *const _, window), None)
+    } else if let (Some(display), Some(surface)) =
+        (window.get_wayland_display(), window.get_wayland_client_surface()) {
+        let (w, h) = window.get_inner_size().unwrap();
+        let egl_surface = wegl::WlEglSurface::new(surface, w as i32, h as i32);
+        (Native::Wayland(display, egl_surface.ptr() as *const _), Some(egl_surface))
+    } else {
+        return Err(CreationError::NotSupported);
+    };
+
+    let context = unsafe {
+        EGLContext::new(native,
+                        attributes,
+                        PixelFormatRequirements {
+                            hardware_accelerated: Some(true),
+                            color_bits: Some(24),
+                            alpha_bits: Some(8),
+                            ..Default::default()
+                        })?
```

Comment:
Okay I have debugged it. This line causes a segfault, if context creation fails (why it fails is a different story, see other comment). The reason is, that the `wl_egl_surface` is dropped, which causes the segfault in this specific case. Any thoughts @vberger ?